payjoin-cli/src/app/v2.rs

```diff
@@ -335,26 +382,128 @@ fn try_contributing_inputs(
+            Some(relay) => relay.clone(),
+            None => return Err(anyhow!("Failed to select from remaining relays")),
+        };
+
+        relay_state.lock().unwrap().set_selected_relay(selected_relay.clone());
+
         let ohttp_keys = {
-            let cert_der = crate::app::read_local_cert()?;
-            payjoin::io::fetch_ohttp_keys_with_cert(ohttp_relay, payjoin_directory, cert_der)
-                .await?
+            payjoin::io::fetch_ohttp_keys(selected_relay.clone(), payjoin_directory.clone()).await
+        };
+
+        match ohttp_keys {
+            Ok(keys) => return Ok(Some(keys)),
+            Err(error) =>
+                if http::StatusCode::from_str(&error.to_string()).is_ok() {
+                    log::debug!("{:?}", error);
+                    return Err(error.into());
+                } else {
+                    log::debug!("Failed to connect to relay: {}, {:?}", selected_relay, error);
+                    relay_state.lock().unwrap().add_failed_relay(selected_relay);
+                },
+        }
+    }
+}
+
+//Local relays are incapable of acting as proxies so we must opportunistically fetch keys from the config
+#[cfg(feature = "_danger-local-https")]
+async fn fetch_keys(
+    config: &Config,
+    _relay_state: Arc<Mutex<RelayState>>,
+) -> Result<Option<payjoin::OhttpKeys>> {
+    let keys = config.v2()?.ohttp_keys.clone().expect("No OHTTP keys set");
+
+    Ok(Some(keys))
+}
+
+#[cfg(not(feature = "_danger-local-https"))]
+async fn validate_relay(
+    config: &Config,
+    relay_state: Arc<Mutex<RelayState>>,
+) -> Result<payjoin::Url> {
+    let payjoin_directory = config.v2()?.pj_directory.clone();
+    let relays = config.v2()?.ohttp_relays.clone();
+
+    loop {
+        let failed_relays = relay_state.lock().unwrap().get_failed_relays();
+
+        let remaining_relays: Vec<_> =
+            relays.iter().filter(|r| !failed_relays.contains(r)).cloned().collect();
+
+        if remaining_relays.is_empty() {
+            return Err(anyhow!("No valid relays available"));
+        }
+
+        let selected_relay = match remaining_relays.choose(&mut rand::thread_rng()) {
+            Some(relay) => relay.clone(),
+            None => return Err(anyhow!("Failed to select from remaining relays")),
         };
-        #[cfg(not(feature = "_danger-local-https"))]
-        let ohttp_keys = payjoin::io::fetch_ohttp_keys(ohttp_relay, payjoin_directory).await?;
-        Ok(ohttp_keys)
+
+        relay_state.lock().unwrap().set_selected_relay(selected_relay.clone());
+
+        let ohttp_keys =
+            payjoin::io::fetch_ohttp_keys(selected_relay.clone(), payjoin_directory.clone()).await;
+
+        match ohttp_keys {
+            Ok(_) => return Ok(selected_relay),
+            Err(error) =>
+                if http::StatusCode::from_str(&error.to_string()).is_ok() {
+                    log::debug!("{:?}", error);
+                    return Err(error.into());
```

Comment:
I added some comparison logic in 8f032e676d1bd8337386d0f0e434e32c0d1855d0 by adding `http` and changing the Error to just be the bare `StatusCode` in